File: solutions/131.Palindrome_Partitioning/AC_dp_nm.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dp_nm.py
# Create Date: 2015-03-19 09:45:12
# Usage:       AC_dp_nm.py 
# Descripton:  


class Solution:
    # @param s, a string
    # @return a list of lists of string
    def partition(self, s):
        slen = len(s)
        indexes = [[] for _ in range(slen)]
        # pre calculate index
        for i in range(slen):
            for j in range(i, slen):
                if s[i: j + 1] == s[i: j + 1][: : -1]:
                    indexes[i].append(j + 1)

        ret = []
        single = []
        def find_partition(start):
            for end in indexes[start]:
                single.append(s[start: end])

                if end == slen:
                    ret.append(single[:])
                else:
                    find_partition(end)

                single.pop()

        find_partition(0)
        return ret


# debug
s = Solution()
print s.partition('aab')

File: solutions/062.Unique_Paths/AC_dp_nm.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dp_nm.cpp
*  Create Date: 2014-12-27 09:07:32
*  Descripton:  dp
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int uniquePaths(int m, int n) {
        // way 1
        // vector<vector<int> > tab(2, vector<int>(n + 1, 1));
        // for (int i = 1; i < m; i++)
        //     for (int j = 1; j < n; j++)
        //         tab[i&1][j] = tab[!(i&1)][j] + tab[i&1][j - 1];
        // return tab[(m-1)&1][n - 1];
        
        vector<int> tab(n, 1);
        for (int i = 1; i < m; i++)
            for (int j = 1; j < n; j++)
                tab[j] += tab[j - 1];
        return tab[n - 1];
    }
};

int main() {
    int n, m;
    Solution s;
    while (cin >> n >> m) {
        cout << s.uniquePaths(n, m);
    }
    return 0;
}

